Sources/Datadog/RUM/RUMMonitor/RUMScope.swift

```diff
@@ -15,3 +15,32 @@ internal protocol RUMScope: class {
     /// * `false` if the scope should be closed.
     func process(command: RUMCommand) -> Bool
 }
+
+extension RUMScope {
+    /// Propagates given `command` to the child scope and manages its lifecycle by
+    /// removing it if it gets closed.
+    func propagate(command: RUMCommand, to chilScope: inout RUMScope?) {
+        if chilScope?.process(command: command) == false {
+            chilScope = nil
+        }
+    }
+
+    /// Propagates given `command` through array of child scopes and manages their lifecycle by
+    /// removing child scopes that get closed.
+    func propagate(command: RUMCommand, to childScopes: inout [RUMScope]) {
+        childScopes = childScopes.compactMap { childScope in
+            let shouldBeKept = childScope.process(command: command)
+            return shouldBeKept ? childScope : nil
+        }
+    }
```

Comment:
name doesn't imply nullifying behavior, can't we achieve this explicitly instead?
```swift
// SessionScope
process(cmd) {
  self.viewScopes = propagate(cmd, viewScopes)
}
// RUMScope
func propagate(cmd, scopes) -> Scopes {
  return scopes.filter { $0.process(cmd) }
}
```

Reply:
This will lead to more code and could be more error-prone if we consider all required ways of owning the scope.

Scopes will be owned in 3 ways:
```swift
var viewScopes: [RUMScope]                 // array
var resourceScopes: [String: RUMScope]     // dictionary keyed by resource name
var actionScope: RUMScope?                 // direct reference
```

With the two `propagate(command:to:)` methods, this is simply:
```swift
propagate(command: command, to: &viewScopes)
propagate(command: command, to: &resourceScopes[resourceName])
propagate(command: command, to: &actionScope)
```

Going "explicit" adds more code code on the caller side:
```swift
viewScopes = propagate(command, to: viewScopes)

if self.resourceScopes[resourceName] != nil {
   resourceScopes = propagate(command, to: resourceScope)
}

self.actionScope = propagate(command, actionScope)
```
and requires 3 different implementations of `propagate(...)` method. Also, when we change to `[RUMViewScope]` https://github.com/DataDog/dd-sdk-ios/pull/190#discussion_r459485890, the `propagate()` method would have to become generic... so it's going even more complex for no reason.

I don't understand what is the problem with current `inout` implementation 🤔. I don't see neither how proposed `propagate(cmd, scopes) -> Scopes` implies the nullifying behaviour vs current one doesn't. 

Would renaming it to `manage(childScopes: inout [RUMScope], byPropagatingCommand command: RUMCommand)` make it better? Please keep in mind - it's just implementation detail of the `RUMScope` that child scopes are closed on propagation. I don't think we need to express it more. LTM WDYT.

Reply:
yes, renaming it to that would be 👍 ✅ 

***

`Dict` case would be `dict[name] = propagate(cmd, dict[name]) // second param is (Scope?)`.
`propagate() -> Scopes` doesn't imply it neither but it doesn't nullify anything, it's the caller who sets things to nil practically and this is the difference.

it can be nice to express when we can even on `private` things as our users may need to read/understand them one day

Reply:
Renamed 👌 here: https://github.com/DataDog/dd-sdk-ios/pull/193/commits/594b5e57c2166e1d676706d028504f962a27604a (to avoid 2PRs rebasing).